File: numl_val_motion_tutorials/src/control_chest_orientation.cpp
#include <ros/ros.h>
#include <ihmc_msgs/ChestTrajectoryRosMessage.h>
#include <ihmc_msgs/SO3TrajectoryPointRosMessage.h>
#include <geometry_msgs/Quaternion.h>
#include <geometry_msgs/Vector3.h>
#include <tf/transform_datatypes.h> // createQuaternionMsgFromRollPitchYaw()
#include <iostream>
#include <math.h>   // M_PI

// Will be used to convert user input to rads
// Rads are used in tf euler coordinates roll,pitch,yaw
double degree2rad(double degree)
{
  return (degree * M_PI)/180;
}

int main(int argc, char* argv[] )
{
  double roll = 0.0, pitch = 0.0, yaw = 0.0;

  // Create a new node to control Valkyrie's chest
  ros::init(argc, argv, "control_chest_orientation");

  // New nodehandler for publishing on a topic
  ros::NodeHandle node;

  // New publisher to publish on the chest_trajectory topic
  ros::Publisher chest_trajectory_pub = node.advertise<ihmc_msgs::ChestTrajectoryRosMessage>
                                          ("/ihmc_ros/valkyrie/control/chest_trajectory", 1);

  // Wait until the publisher is setup, Valkyrie may not respond otherwise
  while (chest_trajectory_pub.getNumSubscribers() == 0) continue;

  // Based on number of variables passed, set the chest to 0 or move to specified coordinates
  if (argc == 1)
  {
    std::cout << "setting to 0 orientation, use 3 parameters for roll, pitch, yaw\n";
  }
  else if (argc == 4)
  {
    roll = degree2rad(std::atof(argv[1]));
    pitch = degree2rad(std::atof(argv[2]));
    yaw = degree2rad(std::atof(argv[3]));
  }
  else if(argc != 4)
  {
    std::cout << "Please enter roll, pitch, and yaw coordinates or nothing for zero chest orientation\n";
    return 0;
  }

  // Transform euler coordinates into quaternions
  geometry_msgs::Quaternion quat = tf::createQuaternionMsgFromRollPitchYaw(roll, pitch, yaw);

  // Create new SO3 Trajectory msg for the Chest trajectory msg
  ihmc_msgs::SO3TrajectoryPointRosMessage SO3_msg;
  SO3_msg.orientation = quat;
  SO3_msg.angular_velocity.x = 0.3;
  SO3_msg.angular_velocity.y = 0.3;
  SO3_msg.angular_velocity.z = 0.3;
  SO3_msg.time = 0.0;

  // Create new Chest Trajectory msg to publish
  ihmc_msgs::ChestTrajectoryRosMessage chest_msg;
  chest_msg.taskspace_trajectory_points.push_back(SO3_msg);
  chest_msg.execution_mode = 0;
  chest_msg.unique_id = 13; // random number; just must not equal zero

  // Publish Chest Trajectory msg
  chest_trajectory_pub.publish(chest_msg);
  sleep(1);

  return 0;
}

File: numl_val_motion_tutorials/scripts/control_chest_orientation.py
#!/usr/bin/env python

import rospy
from ihmc_msgs.msg import *
from geometry_msgs.msg import Quaternion
from geometry_msgs.msg import Vector3
import math
import sys
import tf
import time

if __name__ == '__main__':
    roll = 0.0
    pitch = 0.0
    yaw = 0.0

    # No arguments passed sets chest orientation to 0
    if len(sys.argv) == 1:
        print "Setting to 0 orientation, use 3 parameters for roll, pitch, yaw"

    # Otherwise, assign user inputs to respective trajectories
    elif len(sys.argv) == 4:
        roll = math.radians(float(sys.argv[1]))
        pitch = math.radians(float(sys.argv[2]))
        yaw = math.radians(float(sys.argv[3]))
    elif len(sys.argv) != 4:
        print "Please enter roll, pitch, and yaw coordinates or nothing for zero chest orientation"

    # Create a publisher to publish on the chest_trajectory topic
    chest_trajectory_pub = rospy.Publisher("/ihmc_ros/valkyrie/control/chest_trajectory",
                                           ChestTrajectoryRosMessage, queue_size=1)

    # Create a new node to control the chest
    rospy.init_node("control_chest_orientation")

    # Convert Roll/Pitch/Yaw into Quaternion
    tf_quat = tf.transformations.quaternion_from_euler(roll,pitch,yaw)
    quat = Quaternion(tf_quat[0], tf_quat[1], tf_quat[2], tf_quat[3])

    # Create new msg to publish onto chest_trajectory
    # This msg sets the chest to the 0 position, assuming no args given
    msg = ChestTrajectoryRosMessage(taskspace_trajectory_points=[
                                 SO3TrajectoryPointRosMessage(orientation = quat, angular_velocity =
                                 Vector3(0.3,0.3,0.3), time = 0.0)],
                                 execution_mode = 0, unique_id = 13) # random number; just must no equal zero

    # Publish msg
    chest_trajectory_pub.publish(msg)
    time.sleep(1)
